addons/web/static/src/core/macro.js

```diff
@@ -69,197 +115,64 @@ export class Macro {
-        // Called only once per step.
-        if (!this.stepHasStarted[this.currentIndex]) {
-            delay = this.currentIndex === 0 ? 0 : 50;
-            this.stepHasStarted[this.currentIndex] = true;
-        }
-        return delay;
+    async start() {
+        await this.advance();
     }
 
     async advance() {
-        if (this.isComplete) {
-            return;
-        }
-        if (this.currentStep.trigger) {
-            this.setTimer();
-        }
-        let proceedToAction = true;
-        if (this.currentStep.trigger) {
-            proceedToAction = this.findTrigger();
-        }
-        if (proceedToAction) {
-            this.onStep(this.currentElement, this.currentStep, this.currentIndex);
-            this.clearTimer();
-            const actionResult = await this.stepAction(this.currentElement);
-            if (!actionResult) {
-                // If falsy action result, it means the action worked properly.
-                // So we can proceed to the next step.
-                this.currentIndex++;
-                if (this.currentIndex >= this.steps.length) {
-                    this.stop();
-                }
-                this.debounceAdvance("next");
-            }
-        }
-    }
-
-    /**
-     * Find the trigger and assess whether it can continue on performing the actions.
-     * @returns {boolean}
-     */
-    findTrigger() {
-        const { trigger } = this.currentStep;
-        if (this.isComplete) {
+        if (this.isComplete || this.currentIndex >= this.steps.length) {
+            this.stop();
             return;
         }
         try {
-            if (typeof trigger === "function") {
-                this.currentElement = trigger();
-            } else if (typeof trigger === "string") {
-                const triggerEl = document.querySelector(trigger);
-                this.currentElement = isVisible(triggerEl) && triggerEl;
-            } else {
-                throw new Error(`Trigger can only be string or function.`);
+            const currentStep = this.steps[this.currentIndex];
+            const executeStep = async () => {
+                const trigger = await waitForTrigger(currentStep.trigger);
+                await this.onStep(currentStep, trigger, this.currentIndex);
+                return await performAction(trigger, currentStep.action);
+            };
+            const launchTimer = async () => {
+                const timeout_delay = currentStep.timeout || this.timeout || 10000;
+                await delay(timeout_delay);
+                throw new MacroError(
+                    "Timeout",
+                    `TIMEOUT step failed to complete within ${timeout_delay} ms.`
+                );
+            };
+            // If falsy action result, it means the action worked properly.
+            // So we can proceed to the next step.
+            const actionResult = await Promise.race([executeStep(), launchTimer()]);
```

Comment:
Big change: the timeout now concerns BOTH finding the trigger AND perform the action.
Since the messages that the tours give out include the information of whether or not the trigger has been found, I consider this change acceptable